xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/DBListClass.java

```diff
@@ -282,8 +285,16 @@ public String getQuery(XWikiContext context)
                     select.append("doc." + idField);
                 } else {
                     select.append("idprop.value");
-                    fromStatements.add("StringProperty as idprop");
                     whereStatements.add("obj.id=idprop.id.id and idprop.id.name='" + idField + "'");
+                    // Get the from statements according to the type of property
+                    try {
+                        PropertyClass pc = (PropertyClass) context.getWiki().getDocument(classname, context).getXClass()
+                                .get(idField);
+                        String classType = (pc == null) ? "StringProperty" : pc.newProperty().getClass().getName();
+                        fromStatements.add(classType + " as idprop");
+                    } catch (XWikiException e) {
+                        e.printStackTrace();
```

Comment:
Seems ok. Except that you shouldn't ExceptionUtils.getRootCauseMessage() since we want the stack trace for errors! It should be:

```
LOGGER.error("Failed to get property [{}] for document [{}]", propertyName, className, e);
````
Other comment would be to check if there's a way to access the property class without resorting to calling context.getWiki().getDocument() (haven't checked the code of DBListClass).

Then "we consider that the default type is StringProperty" is not good. You should never explain what you do (that's obvious, the reader just needs to read the code...). You must explain WHY you do it :) So in this case why do you consider that the default type is String Property?